zio/arrowio/reader.go

```diff
@@ -0,0 +1,467 @@
+package arrowio
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+	"unsafe"
+
+	"github.com/apache/arrow/go/v11/arrow"
+	"github.com/apache/arrow/go/v11/arrow/array"
+	"github.com/apache/arrow/go/v11/arrow/ipc"
+	"github.com/apache/arrow/go/v11/parquet/pqarrow"
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/pkg/nano"
+	"github.com/brimdata/zed/zcode"
+	"golang.org/x/exp/slices"
+)
+
+// Reader is a zio.Reader for the Arrow IPC stream format.
+type Reader struct {
+	zctx *zed.Context
+	rr   pqarrow.RecordReader
+
+	typ              zed.Type
+	unionTagMappings map[string][]int
+
+	rec arrow.Record
+	i   int
+
+	builder zcode.Builder
+	val     zed.Value
+}
+
+func NewReader(zctx *zed.Context, r io.Reader) (*Reader, error) {
+	ipcReader, err := ipc.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+	ar, err := NewReaderFromRecordReader(zctx, ipcReader)
+	if err != nil {
+		ipcReader.Release()
+		return nil, err
+	}
+	return ar, nil
+}
+
+func NewReaderFromRecordReader(zctx *zed.Context, rr pqarrow.RecordReader) (*Reader, error) {
+	fields := slices.Clone(rr.Schema().Fields())
+	uniquifyFieldNames(fields)
+	r := &Reader{
+		zctx:             zctx,
+		rr:               rr,
+		unionTagMappings: map[string][]int{},
+	}
+	typ, err := r.newZedType(arrow.StructOf(fields...))
+	if err != nil {
+		return nil, err
+	}
+	r.typ = typ
+	return r, nil
+}
+
+func uniquifyFieldNames(fields []arrow.Field) {
+	names := map[string]int{}
+	for i, f := range fields {
+		if n := names[f.Name]; n > 0 {
+			fields[i].Name += strconv.Itoa(n)
+		}
+		names[f.Name]++
+	}
+}
+
+func (r *Reader) Close() error {
+	if r.rr != nil {
+		r.rr.Release()
+		r.rr = nil
+	}
+	if r.rec != nil {
+		r.rec.Release()
+		r.rec = nil
+	}
+	return nil
+}
+
+func (r *Reader) Read() (*zed.Value, error) {
+	for r.rec == nil {
+		rec, err := r.rr.Read()
+		if err != nil {
+			if err == io.EOF {
+				return nil, nil
+			}
+			return nil, err
+		}
+		if rec.NumRows() > 0 {
+			r.rec = rec
+			r.i = 0
+		} else {
+			rec.Release()
+		}
+	}
+	r.builder.Truncate()
+	for _, array := range r.rec.Columns() {
+		if err := r.buildZcode(array, r.i); err != nil {
+			return nil, err
+		}
+	}
+	r.val = *zed.NewValue(r.typ, r.builder.Bytes())
+	r.i++
+	if r.i >= int(r.rec.NumRows()) {
+		r.rec.Release()
+		r.rec = nil
+	}
+	return &r.val, nil
+}
+
+var dayTimeIntervalFields = []zed.Column{
+	{Name: "days", Type: zed.TypeInt32},
+	{Name: "milliseconds", Type: zed.TypeUint32},
+}
+var decimal128Fields = []zed.Column{
```

Comment:
Let's make an issue for adding decimal128 support.  The type code exists in the spec.

Reply:
@mccanne: There's an existing issue #1522 that speaks to implementing all four of the decimal type variations. We could open separate issues for each if you'd like, or just keep that one.